### ISSUE TITLE
"autoload" triggers on "delete-file-next"

### ISSUE THREAD

**Issue by Qwazzy** [tomasklaen/uosc#924]
Hello. Relatively new user to both MPV and to UOSC, enjoying both immensely. I've been able to figure out most of the things I needed to know via extensive googling and trial-and-error text edits, but this one's stumping me somewhat. (And it's a VERY minor issue, not major, just to be clear up-front.)

I use the player to watch downloaded Youtube videos and delete them afterwards, as a means of removing the need to buffer internet videos and giving me a backlog to work with when an internet outage occurs. So frequent deletion of videos as they're completed is part of my routine.

To that end, I've modified the control bar with an extra button (because I prefer those buttons instead of keybinds or nested menu options):

`command:delete:script-message-to uosc delete-file-next`

It works as intended. The file is sent to the Recycle Bin (as I'm also using `use_trash=yes` in conjunction with this). Then the next file is loaded.

The problem I've run into is when it's also used alongside this:

`autoload=yes`

I like this option too because it means if I'm doing other things while watching videos - for example, building a map for a video game mod on one screen while watching videos on the other - it can automatically load the next video for me when the current one ends. And I don't load the entire folder as a playlist, I use UOSC's built-in file navigator to pick the video I want to watch first, and then autoload will bring up the next video when the first one ends, if I don't delete it first.

The issue is that the two effects stack. So if I'm watching video A and it's about to end, I might mouse over and click that delete button on my control bar. The result is that A is deleted, B is loaded by autoload, and then the "next" part of delete-file-next is called and it skips past B to video C.

This of course is hardly a big issue at all, and I don't mind turning off autoload in order to avoid this behaviour, as it simply means I need to tab over and open a new video when one completes. But if there is a way to make these two features work alongside each other seamlessly, I'm curious to find it. Especially if it's due to me doing something wrong or backwards that can be improved. (I even tried to make some small edits to the code in order to "fix" this for myself, but I'm no coder and I couldn't achieve the desired result.)

I guess the tl;dr here is I'd like some kind of "delete-file" function that doesn't go to next/previous and also doesn't quit the program entirely, which encompasses the three delete-file functions that are currently in UOSC (to my knowledge). I wouldn't mind inserting it locally into my copy alone if need be, I just haven't figured out a way how.

Thanks for your time.

**Comment by Qwazzy** [comment]
For what it's worth, I think I understand what I can do to the code to get my desired result, but I'm being held back by my lack of understanding on LUA and functions. I noticed that when using `delete-file-prev`, the autoload doesn't occur, so it goes back one file and deletes the other. So I tried to modify the main.lua to have that function also do `navigate_item(1)` or `navigate_directory(1)`, but that second action doesn't occur. I must be misunderstanding how to have a function do more than one thing.

**Comment by christoph-heinrich** [comment]
You can combine both commands on one button by separating them with a `;`.
So in this case `command:delete:script-message-to uosc delete-file-prev;script-message-to uosc next-file`.
I didn't test it for this specifically, but that generally works.

My preferred script for deleting files is [delete_file.lua](https://github.com/zenyd/mpv-scripts/blob/master/delete_file.lua), in case the above doesn't work.

**Comment by tomasklaen** [comment]
The sequence in `delete-fine-next` is:

1. Start playing next file.
2. Delete previous one.

Not sure atm why that would be triggering autoload. Unless switching file triggers some `file-end` event on current one. In which case we'd need to figure out how to filter that out... :/

**Comment by Qwazzy** [comment]
> You can combine both commands on one button by separating them with a ;.
So in this case command:delete:script-message-to uosc delete-file-prev;script-message-to uosc next-file.
I didn't test it for this specifically, but that generally works.

I've seen an example of this elsewhere so you are probably correct, however sadly it doesn't seem to work for these two commands. Perhaps it has something to do with them trying to execute near-simultaneously. I don't suppose there's any kind of "wait" or "delay" command? If I could space them about a second apart, I bet it would work. Kinda like how my current solution of two buttons works fine, since I'm not clicking them at the same time, there's always a delay between moving the mouse from one to the other.

> My preferred script for deleting files is [delete_file.lua](https://github.com/zenyd/mpv-scripts/blob/master/delete_file.lua), in case the above doesn't work.

I was considering this one as well, but the documentation didn't explain very well if this script can move files to the Recycle Bin instead of deleting them permanently. Do you have any experience with that? If it can, I'd definitely be willing to use this as a replacement for my current method.

> Not sure atm why that would be triggering autoload. Unless switching file triggers some file-end event on current one. In which case we'd need to figure out how to filter that out... :/

I wish I knew. Using UOSC's control bar to go to the next file doesn't make it trigger twice, I've only noticed this effect specifically when using `delete-file-next`. I've looked at the code for the `delete_file_navigate(delta)` command to see if I could figure out how it works and maybe alter it or copy it to a new function for my purpose, but I just can't wrap my head around it.

**Comment by christoph-heinrich** [comment]
> I was considering this one as well, but the documentation didn't explain very well if this script can move files to the Recycle Bin instead of deleting them permanently.

It doesn't support that afaik, but it can move files to a specific directory.

**Comment by Qwazzy** [comment]
I'm gonna go ahead a close this for now - while I didn't obtain the solution I was aiming for, I'm not far off from it, and the secondary goal of bringing this bug / unintended side effect / whatever you'd like to call it, has been achieved as well. It's odd behaviour, but not significantly so.

I appreciate the replies, and if anyone has further thoughts to share they're of course welcome to do so 👍

**Comment by tomasklaen** [comment]
This is a valid issue that _should_ be fixed, so keep this open until it is.

**Comment by Qwazzy** [comment]
I wanted to make sure this was reproducible, so I downloaded a fresh copy of MPV for Windows (mpv-x86_64-v3-20240623-git-265056f) and UOSC 5.2.0 and set it up in what I believe is the same process I set up my first instance, and tried the `delete-file-next' command again, and... the bug didn't occur, it only went forwards once as it should.

So I went back to my actual daily-use copy and redid the control bar button, and... again, no bug. Worked as intended.

There might still be an unintended bug in here somewhere, but this is making me believe now that it's more likely an issue specifically on my end and not in UOSC as a whole. 

Either that, or my PC is gaslighting me. Wouldn't be the first time, honestly...
